sdk/go/pulumi/context.go

```diff
@@ -177,15 +183,21 @@ func (ctx *Context) Context() context.Context {
 	return ctx.ctx
 }
 
+// Returns a copy of base context in which the value associated with key is val.
+func (ctx *Context) WithValue(key, val any) context.Context {
+	ctx.ctx = context.WithValue(ctx.ctx, key, val)
```

Comment:
This should make a copy of `ctx` and then assign the new WithValue context to that and return the new overall `pulumi.Context`

Reply:
Gotcha. Made a copy of whole pulumi context and update the Go's one with new key val. I have a test that confirm that old context is untouched and states are the same.

Reply:
I was going back and forth a bit here wondering why we should be making the copy of  `ctx` here, but in the end I think I agree with it.  I think a small downside to that is that the user now needs to think about what to do with closing that context, etc., but for passing the context down it would be really unexpected if the key value pairs were suddenly available in a caller of a function, if that function calls `ctx.WithValue()`.  So making a copy makes sense to me.